### DISCUSSION TITLE
Local AWS Lambda hosting in Aspire

### DISCUSSION THREAD

**Discussion by Kralizek** [microsoft/aspire#5692] [Show and tell]
The juicy bits first.

```csharp
var builder = DistributedApplication.CreateBuilder(args);

var lambda = builder.AddLambdaLocalHosting("lambda")
    .WithEndpoint(scheme: "http", port: 52000);

lambda.AddAnnotationFunction(typeof(Functions), nameof(Functions.GetCountriesAsync));

builder.AddProject<Projects.FindNationalityWeb>("web")
    .WithReference(lambda)
    .WithEnvironment("AWS_ACCESS_KEY_ID", "fake")
    .WithEnvironment("AWS_SECRET_ACCESS_KEY", "fake")
    .WithExternalHttpEndpoints();

builder.Build().Run();
```

I was working on this before the summer. In the last couple of days I decided to take the different tests I wrote back then and publish in a nice soon-to-be NuGet package.

Behind the scenes, I use the same code used for YARP to host in memory a ASP.NET Core application. The application has a [single endpoint](https://github.com/Kralizek/dotnet-aspire-lambda-hosting/blob/63d28dfccc46ccaca7f2bd9950db88ae4c61a701/src/AWSLambda/LambdaResourceLifecycleHook.cs#L25) that matches the AWS API to invoke a function execution via their SDKs. The endpoint plays some [crazy (and quite dirty, truth to be told) reflection tricks](https://github.com/Kralizek/dotnet-aspire-lambda-hosting/blob/63d28dfccc46ccaca7f2bd9950db88ae4c61a701/src/AWSLambda/LambdaEndpoint.cs#L11-L59) to instantiate and execute the function. The reflection stuff would be so much nicer/easier/neat if https://github.com/aws/aws-lambda-dotnet/issues/1760 were to be solved.

Right now I support only function created with Annotation framework, but ideally classic function programming models would be supported.

What else? Oh, the most important thing! The calling application will never know anything about Aspire hosting. The [`WithReference`](https://github.com/Kralizek/dotnet-aspire-lambda-hosting/blob/63d28dfccc46ccaca7f2bd9950db88ae4c61a701/src/AWSLambda/LambdaHostingExtensions.cs#L27-L32) extension used in the snippet above is a custom one setting the SDK local endpoint to the in-memory web app address. 

```csharp
public static IResourceBuilder<TResource> WithReference<TResource>(this IResourceBuilder<TResource> builder, IResourceBuilder<LambdaResource> lambda) where TResource : IResourceWithEnvironment
{
    builder.WithEnvironment("AWS_ENDPOINT_URL_Lambda", lambda.GetEndpoint("http"));

    return builder;
}
```

Here is the link to the repository: https://github.com/Kralizek/dotnet-aspire-lambda-hosting

**Comment by En3Tho** [comment]
I wish I had something like this when I worked with lambdas :)

**Reply by Kralizek** [reply]
I know your pain. That's exactly why I start juggling with this :D

Years ago, I wrote [this one](https://github.com/keghub/lambda-local-runner) but it required heavy modification on the client that never made me really happy.
